File: DualKeyEncryption.vb
Imports System.Numerics
Imports System.Globalization

Public Class DualKeyEncryption
    ''
    ' STEP VI:
    ''
    Public Function encodeWord(msg As Char(), key As BigInteger, semi_prime As BigInteger) As String()
        Dim encryption(msg.Length - 1) As String
        Dim x As Integer
        For i As Integer = 0 To msg.Length - 1
            ' get unicode of this character as x
            x = AscW(msg(i))
            ' use RSA to encrypt & save in base 16
            encryption(i) = BigInteger.ModPow(x, key, semi_prime).ToString("X")
        Next i

        Return encryption
    End Function

    ''
    ' STEP VII:
    ''
    Public Function decodeWord(code As String(), key As BigInteger, semi_prime As BigInteger) As String
        Dim decryption As String = ""
        Dim c As BigInteger
        For i As Integer = 0 To code.Length - 1
            ' use RSA to decrypt
            c = BigInteger.ModPow(BigInteger.Parse(code(i), NumberStyles.HexNumber), key, semi_prime)
            decryption &= ChrW(c)
        Next i

        Return decryption
    End Function

End Class

File: DualKeyEncryption.cs
using System;
using System.Numerics;
using System.Globalization;

namespace Miscellaneous_CS
{
    class DualKeyEncryption
    {

        BigInteger semi_prime;

        public DualKeyEncryption(BigInteger semi_prime)
        {
            this.semi_prime = semi_prime;
        }

        /*
         * STEP VI:
         */
        public string[] encodeWord(char[] msg, BigInteger key)
        {
            string[] encryption = new string[msg.Length];
            int x;
            for (int i = 0; i < msg.Length; i++)
            {
                // get unicode of this character as x
                x = (int)msg[i];
                // use RSA to encrypt & save in base 16
                encryption[i] = BigInteger.ModPow(x, key, semi_prime).ToString("X");
            }

            return encryption;
        }

        /*
         * STEP VII:
         */
        public string decodeWord(string[] code, BigInteger key)
        {
            String decryption = "";
            BigInteger c;
            for (int i = 0; i < code.Length; i++)
            {
                // use RSA to decrypt
                c = BigInteger.ModPow(BigInteger.Parse(code[i], NumberStyles.HexNumber), key, semi_prime);
                decryption += (char)c;
            }

            return decryption;
        }
    }
}
